src/main/java/com/google/api/gax/grpc/PageStreamingCallable.java

```diff
@@ -60,62 +65,120 @@ public String toString() {
-          return false;
-        }
-        ResponseT newPage =
-            Futures.getUnchecked(callable.futureCall(context.withRequest(nextRequest)));
-        currentIter = pageDescriptor.extractResources(newPage).iterator();
-
-        Object nextToken = pageDescriptor.extractNextToken(newPage);
-        if (nextToken.equals(pageDescriptor.emptyToken())) {
-          nextRequest = null;
+      protected ResourceT computeNext() {
+        if (currentIterator.hasNext()) {
+          return currentIterator.next();
+        } else if (nextRequest == null) {
+          return endOfData();
         } else {
-          nextRequest = pageDescriptor.injectToken(nextRequest, nextToken);
+          try {
+            ResponseT newPage =
+                Futures.getUnchecked(callable.futureCall(context.withRequest(nextRequest)));
+            Object nextToken = pageDescriptor.extractNextToken(newPage);
+            if (nextToken.equals(pageDescriptor.emptyToken())) {
+              nextRequest = null;
+            } else {
+              nextRequest = pageDescriptor.injectToken(nextRequest, nextToken);
+            }
+            currentIterator = pageDescriptor.extractResources(newPage).iterator();
+            return computeNext();
+          } catch (UncheckedExecutionException e) {
+            handleException(e);
+            return null;
+          }
         }
-        return currentIter.hasNext();
       }
+    }
 
-      @Override
-      public ResourceT next() {
-        if (!hasNext()) {
-          throw new NoSuchElementException();
+    @Override
+    public Iterable<ResourceT> getPageValues() {
+      return pageDescriptor.extractResources(getPage());
+    }
+
+    @Override
+    public PageAccessor<ResourceT> getNextPage() {
+      Object nextToken = getNextPageToken();
+      if (nextToken == null) {
+        return null;
+      } else {
+        RequestT nextRequest =
+            pageDescriptor.injectToken(context.getRequest(), getNextPageToken());
+        return new PageAccessorImpl<>(callable,
+                                      pageDescriptor,
+                                      context.withRequest(nextRequest));
+      }
+    }
+
+    @Override
+    public String getNextPageToken() {
+      Object nextToken = pageDescriptor.extractNextToken(getPage());
+      if (nextToken == null || nextToken.equals(pageDescriptor.emptyToken())) {
+        return null;
+      } else {
+        return nextToken.toString();
+      }
+    }
+
+    private ResponseT getPage() {
+      if (currentPage == null) {
+        try {
+          currentPage = Futures.getUnchecked(callable.futureCall(context));
+        } catch (UncheckedExecutionException e) {
+          handleException(e);
         }
-        return currentIter.next();
       }
+      return currentPage;
+    }
 
-      @Override
-      public void remove() {
-        throw new UnsupportedOperationException();
+    private void handleException(UncheckedExecutionException exception) {
```

Comment:
I think you will be better off if you call this method `getUnchecked` passing to it the `Callable` and use `Futures.getUnchecked` in the method (rather than assuming that you are handling an exception that is thrown
by a usage of `Futures.getUnchecked`).